### ISSUE TITLE
Problem with Flent Average for rrul & rrul_be?

### ISSUE THREAD

**Issue by richb-hanover** [tohojo/flent#184]
Flent 1.9.9-git-2a08dd9:

I am confused about the following. Might they be a problem?

* The displayed UDP ping times in the Average column for the rrul and rrul_be tests don't seem correct - they are considerably lower than expected. (The median values are believable.) 

* The plotted Average ping for rrul & rrul_be are so much lower than the readings. (see plots below)

Thanks.
```
Started Flent 1.9.9-git-2a08dd9 using Python 3.7.3.
Starting rrul test. Expected run time: 70 seconds.
Data file written to ./rrul-2019-09-15T065804.053659.Richs-MBP-10337_lan.flent.gz.

Summary of rrul test run from 2019-09-15 10:58:04.053659
  Title: 'Richs-MBP-10337.lan'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       106.25       105.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        31.23        30.55 ms              350
 Ping (ms) UDP BE               :         9.43       106.38 ms              352
 Ping (ms) UDP BK               :         9.40       106.72 ms              352
 Ping (ms) UDP EF               :         9.38       106.44 ms              352
 Ping (ms) avg                  :         9.40          N/A ms              352
 TCP download BE                :         1.52         1.55 Mbits/s         352
 TCP download BK                :         1.52         1.52 Mbits/s         352
 TCP download CS5               :         1.51         1.50 Mbits/s         352
 TCP download EF                :         1.47         1.53 Mbits/s         352
 TCP download avg               :         1.50          N/A Mbits/s         352
 TCP download sum               :         6.02          N/A Mbits/s         352
 TCP totals                     :         6.45          N/A Mbits/s         352
 TCP upload BE                  :         0.10         0.10 Mbits/s         352
 TCP upload BK                  :         0.11         0.11 Mbits/s         352
 TCP upload CS5                 :         0.11         0.10 Mbits/s         352
 TCP upload EF                  :         0.11         0.11 Mbits/s         352
 TCP upload avg                 :         0.11          N/A Mbits/s         352
 TCP upload sum                 :         0.43          N/A Mbits/s         352
Started Flent 1.9.9-git-2a08dd9 using Python 3.7.3.
Starting rrul_be test. Expected run time: 70 seconds.
Data file written to ./rrul_be-2019-09-15T065924.850091.Richs-MBP-10337_lan.flent.gz.

Summary of rrul_be test run from 2019-09-15 10:59:24.850091
  Title: 'Richs-MBP-10337.lan'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       106.67       106.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        30.96        30.30 ms              350
 Ping (ms) UDP BE1              :         9.47       106.33 ms              352
 Ping (ms) UDP BE2              :         9.40       107.30 ms              352
 Ping (ms) UDP BE3              :         9.26       104.88 ms              352
 Ping (ms) avg                  :         9.38          N/A ms              352
 TCP download BE                :         1.38         1.39 Mbits/s         352
 TCP download BE2               :         1.61         1.67 Mbits/s         352
 TCP download BE3               :         1.55         1.61 Mbits/s         352
 TCP download BE4               :         1.50         1.56 Mbits/s         352
 TCP download avg               :         1.51          N/A Mbits/s         352
 TCP download sum               :         6.04          N/A Mbits/s         352
 TCP totals                     :         6.48          N/A Mbits/s         352
 TCP upload BE                  :         0.11         0.11 Mbits/s         352
 TCP upload BE2                 :         0.11         0.11 Mbits/s         352
 TCP upload BE3                 :         0.11         0.11 Mbits/s         352
 TCP upload BE4                 :         0.11         0.10 Mbits/s         352
 TCP upload avg                 :         0.11          N/A Mbits/s         352
 TCP upload sum                 :         0.44          N/A Mbits/s         352
bash-3.2$
```
<img width="819" alt="image" src="https://user-images.githubusercontent.com/1094930/64921053-35d42680-d78c-11e9-990d-4d57b92ec379.png">

<img width="819" alt="image" src="https://user-images.githubusercontent.com/1094930/64921056-3ff62500-d78c-11e9-8cfa-cd63f91861c4.png">

**Comment by tohojo** [comment]
Hmm, hard to say for sure without data and log files, but this might be an artifact of the netperf UDP RR test. I would expect you'd get better data using irtt...

**Comment by richb-hanover** [comment]
OK New run, this time with result and log files... (Charts are substantially the same as above.) Thanks!

[rrul_be-2019-09-15T125209.510692.Richs-MBP-10337_lan.log](https://github.com/tohojo/flent/files/3614011/rrul_be-2019-09-15T125209.510692.Richs-MBP-10337_lan.log)
[rrul_be-2019-09-15T125209.510692.Richs-MBP-10337_lan.flent.gz](https://github.com/tohojo/flent/files/3614012/rrul_be-2019-09-15T125209.510692.Richs-MBP-10337_lan.flent.gz)
[rrul-2019-09-15T125048.680495.Richs-MBP-10337_lan.log](https://github.com/tohojo/flent/files/3614013/rrul-2019-09-15T125048.680495.Richs-MBP-10337_lan.log)
[rrul-2019-09-15T125048.680495.Richs-MBP-10337_lan.flent.gz](https://github.com/tohojo/flent/files/3614014/rrul-2019-09-15T125048.680495.Richs-MBP-10337_lan.flent.gz)

```
Started Flent 1.9.9-git-436d5b6 using Python 3.7.3.
Starting rrul test. Expected run time: 70 seconds.
Data file written to ./rrul-2019-09-15T125048.680495.Richs-MBP-10337_lan.flent.gz

Summary of rrul test run from 2019-09-15 16:50:48.680495
  Title: 'Richs-MBP-10337.lan'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       106.26       105.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        30.82        30.25 ms              350
 Ping (ms) UDP BE               :         9.47       105.93 ms              353
 Ping (ms) UDP BK               :         9.11       106.72 ms              353
 Ping (ms) UDP EF               :         9.43       105.93 ms              353
 Ping (ms) avg                  :         9.34          N/A ms              353
 TCP download BE                :         1.60         1.66 Mbits/s         353
 TCP download BK                :         1.49         1.53 Mbits/s         353
 TCP download CS5               :         1.47         1.53 Mbits/s         353
 TCP download EF                :         1.44         1.46 Mbits/s         353
 TCP download avg               :         1.50          N/A Mbits/s         353
 TCP download sum               :         6.00          N/A Mbits/s         353
 TCP totals                     :         6.42          N/A Mbits/s         353
 TCP upload BE                  :         0.10         0.10 Mbits/s         353
 TCP upload BK                  :         0.10         0.10 Mbits/s         353
 TCP upload CS5                 :         0.11         0.11 Mbits/s         353
 TCP upload EF                  :         0.11         0.11 Mbits/s         353
 TCP upload avg                 :         0.11          N/A Mbits/s         353
 TCP upload sum                 :         0.42          N/A Mbits/s         353
Started Flent 1.9.9-git-436d5b6 using Python 3.7.3.
Starting rrul_be test. Expected run time: 70 seconds.
Data file written to ./rrul_be-2019-09-15T125209.510692.Richs-MBP-10337_lan.flent.gz

Summary of rrul_be test run from 2019-09-15 16:52:09.510692
  Title: 'Richs-MBP-10337.lan'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       106.03       105.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        31.53        31.20 ms              350
 Ping (ms) UDP BE1              :         9.23       105.43 ms              353
 Ping (ms) UDP BE2              :         9.44       106.72 ms              353
 Ping (ms) UDP BE3              :         9.53       104.99 ms              353
 Ping (ms) avg                  :         9.40          N/A ms              353
 TCP download BE                :         1.54         1.57 Mbits/s         353
 TCP download BE2               :         1.43         1.46 Mbits/s         353
 TCP download BE3               :         1.59         1.67 Mbits/s         353
 TCP download BE4               :         1.51         1.55 Mbits/s         353
 TCP download avg               :         1.52          N/A Mbits/s         353
 TCP download sum               :         6.07          N/A Mbits/s         353
 TCP totals                     :         6.50          N/A Mbits/s         353
 TCP upload BE                  :         0.11         0.11 Mbits/s         353
 TCP upload BE2                 :         0.10         0.11 Mbits/s         353
 TCP upload BE3                 :         0.11         0.11 Mbits/s         353
 TCP upload BE4                 :         0.11         0.10 Mbits/s         353
 TCP upload avg                 :         0.11          N/A Mbits/s         353
 TCP upload sum                 :         0.43          N/A Mbits/s         353
```

**Comment by tohojo** [comment]
Rich Brown <notifications@github.com> writes:

> OK New run, this time with result and log files... (Charts are
> substantially the same as above.) Thanks!

Aha! That's useful!

Right, so there are two things going on here:

- The MEAN_VALUE stored as metadata has not been transformed from
  transactions/second to ms.

- The extra ping test was included in the average but not on the plot.

Both of these should be fixed in current master, although the former
will only work for new data files. Not sure if I can find a good way to
detect that the transformation is missing and fix it in post-processing,
unfortunately :(

**Comment by richb-hanover** [comment]
Looks good. Displayed ping averages are much more reasonable. 

Does the plotted Ping Average include/average in the extra host (1.1.1.1)? Is this new? Thanks.

```
Started Flent 1.9.9-git-83a0308 using Python 3.7.3.
Starting rrul test. Expected run time: 70 seconds.
Data file written to ./rrul-2019-09-16T190433.498468.Rich_s_Mac.flent.gz

Summary of rrul test run from 2019-09-16 23:04:33.498468
  Title: 'Rich's_Mac'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       107.09       106.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        31.74        31.95 ms              350
 Ping (ms) UDP BE               :       104.60       104.82 ms              352
 Ping (ms) UDP BK               :       104.60       104.93 ms              352
 Ping (ms) UDP EF               :       103.52       103.95 ms              352
 Ping (ms) avg                  :       104.24          N/A ms              352
 TCP download BE                :         1.40         1.41 Mbits/s         352
 TCP download BK                :         1.46         1.50 Mbits/s         352
 TCP download CS5               :         1.78         1.83 Mbits/s         352
 TCP download EF                :         1.53         1.55 Mbits/s         352
 TCP download avg               :         1.54          N/A Mbits/s         352
 TCP download sum               :         6.17          N/A Mbits/s         352
 TCP totals                     :         6.64          N/A Mbits/s         352
 TCP upload BE                  :         0.12         0.11 Mbits/s         352
 TCP upload BK                  :         0.12         0.11 Mbits/s         352
 TCP upload CS5                 :         0.11         0.11 Mbits/s         352
 TCP upload EF                  :         0.12         0.11 Mbits/s         352
 TCP upload avg                 :         0.12          N/A Mbits/s         352
 TCP upload sum                 :         0.47          N/A Mbits/s         352
Started Flent 1.9.9-git-83a0308 using Python 3.7.3.
Starting rrul_be test. Expected run time: 70 seconds.
Data file written to ./rrul_be-2019-09-16T190554.907631.Rich_s_Mac.flent.gz

Summary of rrul_be test run from 2019-09-16 23:05:54.907631
  Title: 'Rich's_Mac'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       105.73       104.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        31.27        30.90 ms              349
 Ping (ms) UDP BE1              :       104.28       104.38 ms              352
 Ping (ms) UDP BE2              :       104.06       103.84 ms              352
 Ping (ms) UDP BE3              :       105.60       105.60 ms              352
 Ping (ms) avg                  :       104.64          N/A ms              352
 TCP download BE                :         1.46         1.47 Mbits/s         352
 TCP download BE2               :         1.52         1.56 Mbits/s         352
 TCP download BE3               :         1.53         1.56 Mbits/s         352
 TCP download BE4               :         1.64         1.70 Mbits/s         352
 TCP download avg               :         1.54          N/A Mbits/s         352
 TCP download sum               :         6.15          N/A Mbits/s         352
 TCP totals                     :         6.61          N/A Mbits/s         352
 TCP upload BE                  :         0.10         0.10 Mbits/s         352
 TCP upload BE2                 :         0.12         0.11 Mbits/s         352
 TCP upload BE3                 :         0.12         0.11 Mbits/s         352
 TCP upload BE4                 :         0.12         0.11 Mbits/s         352
 TCP upload avg                 :         0.11          N/A Mbits/s         352
 TCP upload sum                 :         0.46          N/A Mbits/s         352
```
<img width="1171" alt="image" src="https://user-images.githubusercontent.com/1094930/64999472-ec7ff600-d8b5-11e9-963e-c5aa353756d3.png">

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/1094930/64999484-f86bb800-d8b5-11e9-8961-0f0b2f0e5883.png">

**Comment by tohojo** [comment]
Rich Brown <notifications@github.com> writes:

> Looks good. Displayed ping averages are much more reasonable.

Awesome!

> Does the plotted Ping Average include/average in the extra host
> (1.1.1.1)? Is this new? Thanks.

Yeah, it does. And no, it isn't. I'm wondering why it isn't in the
summary output...

**Comment by flent-users** [comment]
Toke Høiland-Jørgensen <notifications@github.com> writes:

> Rich Brown <notifications@github.com> writes:
>
>> Looks good. Displayed ping averages are much more reasonable.
>
> Awesome!
>
>> Does the plotted Ping Average include/average in the extra host
>> (1.1.1.1)? Is this new? Thanks.
>
> Yeah, it does. And no, it isn't. I'm wondering why it isn't in the
> summary output...

Right, so the reason for this is that in your tests, fping is being
killed by the watchdog timer, which uses SIGTERM, and thus prevents it
from writing its end-of-test averages. Let's fix this so the averages
get accounted properly...

**Comment by richb-hanover** [comment]
I'm not sure I see any difference... I attach the flent.gz and log files.

[Rich_Test_gz_&_logs.zip](https://github.com/tohojo/flent/files/3631723/Rich_Test_gz_._logs.zip)


<img width="1009" alt="image" src="https://user-images.githubusercontent.com/1094930/65253161-2c1e2c00-dac8-11e9-88cc-bdb23b625989.png">

```
Started Flent 1.9.9-git-67ac0f7 using Python 3.7.3.
Starting rrul test. Expected run time: 70 seconds.
Data file written to ./rrul-2019-09-19T102204.337155.Rich_Test.flent.gz

Summary of rrul test run from 2019-09-19 14:22:04.337155
  Title: 'Rich_Test'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       107.87       107.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        30.44        30.20 ms              350
 Ping (ms) UDP BE               :       103.84       104.17 ms              354
 Ping (ms) UDP BK               :       104.28       104.17 ms              354
 Ping (ms) UDP EF               :       104.28       104.49 ms              354
 Ping (ms) avg                  :       104.13          N/A ms              354
 TCP download BE                :         1.35         1.39 Mbits/s         354
 TCP download BK                :         1.51         1.59 Mbits/s         354
 TCP download CS5               :         1.47         1.49 Mbits/s         354
 TCP download EF                :         1.82         1.85 Mbits/s         354
 TCP download avg               :         1.54          N/A Mbits/s         354
 TCP download sum               :         6.15          N/A Mbits/s         354
 TCP totals                     :         6.62          N/A Mbits/s         354
 TCP upload BE                  :         0.12         0.11 Mbits/s         354
 TCP upload BK                  :         0.12         0.11 Mbits/s         354
 TCP upload CS5                 :         0.12         0.11 Mbits/s         354
 TCP upload EF                  :         0.11         0.11 Mbits/s         354
 TCP upload avg                 :         0.12          N/A Mbits/s         354
 TCP upload sum                 :         0.47          N/A Mbits/s         354
Started Flent 1.9.9-git-67ac0f7 using Python 3.7.3.
Starting rrul_be test. Expected run time: 70 seconds.
Data file written to ./rrul_be-2019-09-19T102325.892621.Rich_Test.flent.gz

Summary of rrul_be test run from 2019-09-19 14:23:25.892621
  Title: 'Rich_Test'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       106.41       105.00 ms              350
 Ping (ms) ICMP 1.1.1.1 (extra) :        30.66        30.45 ms              350
 Ping (ms) UDP BE1              :       104.60       104.38 ms              354
 Ping (ms) UDP BE2              :       104.93       104.88 ms              354
 Ping (ms) UDP BE3              :       103.73       103.84 ms              354
 Ping (ms) avg                  :       104.42          N/A ms              354
 TCP download BE                :         1.64         1.70 Mbits/s         354
 TCP download BE2               :         1.54         1.60 Mbits/s         354
 TCP download BE3               :         1.43         1.48 Mbits/s         354
 TCP download BE4               :         1.57         1.62 Mbits/s         354
 TCP download avg               :         1.54          N/A Mbits/s         354
 TCP download sum               :         6.18          N/A Mbits/s         354
 TCP totals                     :         6.62          N/A Mbits/s         354
 TCP upload BE                  :         0.11         0.11 Mbits/s         354
 TCP upload BE2                 :         0.11         0.11 Mbits/s         354
 TCP upload BE3                 :         0.11         0.11 Mbits/s         354
 TCP upload BE4                 :         0.11         0.11 Mbits/s         354
 TCP upload avg                 :         0.11          N/A Mbits/s         354
 TCP upload sum                 :         0.44          N/A Mbits/s         354
```

**Comment by tohojo** [comment]
Rich Brown <notifications@github.com> writes:

> I'm not sure I see any difference... I attach the flent.gz and log
> files.

Huh, no, seems I was wrong in my root cause analysis. The issue is that
it's actually being output on stderr, not stdout.

Should be fixed in fa47b0f6df7de6cc35eca5b6d04f77c2c59bdbf1...

**Comment by richb-hanover** [comment]
Better. Thanks!

```
Started Flent 1.9.9-git-fa47b0f using Python 3.7.3.
Starting rrul test. Expected run time: 70 seconds.
Data file written to ./rrul-2019-09-19T123807.165015.Rich_Test_-_Flent_1_9_9-git-fa47b0f.flent.gz

Summary of rrul test run from 2019-09-19 16:38:07.165015
  Title: 'Rich_Test_-_Flent_1.9.9-git-fa47b0f'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       107.00       105.00 ms              354
 Ping (ms) ICMP 1.1.1.1 (extra) :        30.00        29.20 ms              354
 Ping (ms) UDP BE               :       106.84       107.41 ms              354
 Ping (ms) UDP BK               :       106.84       106.95 ms              354
 Ping (ms) UDP EF               :       106.38       107.07 ms              354
 Ping (ms) avg                  :        91.41          N/A ms              354
 TCP download BE                :         1.50         1.52 Mbits/s         354
 TCP download BK                :         1.54         1.56 Mbits/s         354
 TCP download CS5               :         1.43         1.48 Mbits/s         354
 TCP download EF                :         1.59         1.68 Mbits/s         354
 TCP download avg               :         1.52          N/A Mbits/s         354
 TCP download sum               :         6.06          N/A Mbits/s         354
 TCP totals                     :         6.50          N/A Mbits/s         354
 TCP upload BE                  :         0.10         0.10 Mbits/s         354
 TCP upload BK                  :         0.11         0.11 Mbits/s         354
 TCP upload CS5                 :         0.11         0.11 Mbits/s         354
 TCP upload EF                  :         0.12         0.11 Mbits/s         354
 TCP upload avg                 :         0.11          N/A Mbits/s         354
 TCP upload sum                 :         0.44          N/A Mbits/s         354
Started Flent 1.9.9-git-fa47b0f using Python 3.7.3.
Starting rrul_be test. Expected run time: 70 seconds.
Data file written to ./rrul_be-2019-09-19T123928.602131.Rich_Test_-_Flent_1_9_9-git-fa47b0f.flent.gz

Summary of rrul_be test run from 2019-09-19 16:39:28.602131
  Title: 'Rich_Test_-_Flent_1.9.9-git-fa47b0f'

                                           avg       median          # data pts
 Ping (ms) ICMP                 :       106.00       105.00 ms              355
 Ping (ms) ICMP 1.1.1.1 (extra) :        30.70        29.65 ms              355
 Ping (ms) UDP BE1              :       105.26       105.71 ms              355
 Ping (ms) UDP BE2              :       105.26       105.60 ms              355
 Ping (ms) UDP BE3              :       105.60       105.54 ms              355
 Ping (ms) avg                  :        90.56          N/A ms              355
 TCP download BE                :         1.51         1.55 Mbits/s         355
 TCP download BE2               :         1.67         1.67 Mbits/s         355
 TCP download BE3               :         1.42         1.42 Mbits/s         355
 TCP download BE4               :         1.40         1.45 Mbits/s         355
 TCP download avg               :         1.50          N/A Mbits/s         355
 TCP download sum               :         6.00          N/A Mbits/s         355
 TCP totals                     :         6.44          N/A Mbits/s         355
 TCP upload BE                  :         0.11         0.11 Mbits/s         355
 TCP upload BE2                 :         0.11         0.11 Mbits/s         355
 TCP upload BE3                 :         0.11         0.11 Mbits/s         355
 TCP upload BE4                 :         0.11         0.11 Mbits/s         355
 TCP upload avg                 :         0.11          N/A Mbits/s         355
 TCP upload sum                 :         0.44          N/A Mbits/s         355
```
